langchain/google_search.py

```diff
@@ -0,0 +1,135 @@
+    """Wrapper for Google Search API.
+
+    TODO: DOCS for using it
+    1. Install google-api-python-client
+    - If you don't already have a Google account, sign up.
+    - If you have never created a Google APIs Console project,
+    read the Managing Projects page and create a project in the Google API Console.
+    - Install the library using pip install google-api-python-client
+    The current version of the library is 2.70.0 at this time
+
+    2. To create an API key:
+    - Navigate to the APIs & Services→Credentials panel in Cloud Console.
+    - Select Create credentials, then select API key from the drop-down menu.
+    - The API key created dialog box displays your newly created key.
+    - You now have an API_KEY
+
+    3. Setup Custom Search Engine so you can search the entire web
+    - Create a custom search engine in this link.
+    - In Sites to search, add any valid URL (i.e. www.stackoverflow.com).
+    - That’s all you have to fill up, the rest doesn’t matter.
+    In the left-side menu, click Edit search engine → {your search engine name}
+    → Setup Set Search the entire web to ON. Remove the URL you added from
+     the list of Sites to search.
+    - Under Search engine ID you’ll find the search-engine-ID.
+
+    4. Enable the Custom Search API
+    - Navigate to the APIs & Services→Dashboard panel in Cloud Console.
+    - Click Enable APIs and Services.
+    - Search for Custom Search API and click on it.
+    - Click Enable.
+    URL for it: https://console.cloud.google.com/apis/library/customsearch.googleapis
+    .com
+    Adapted from: Instructions adapated from https://stackoverflow.com/questions/
+    37083058/
+    programmatically-searching-google-in-python-using-custom-search
+
+
+    """
+
+    search_engine: Any  #: :meta private:
+
+    google_api_key: Optional[str] = None
+    google_cse_id: Optional[str] = None
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+
+    def _google_search_results(
+        self,
+        search_term: str,
+        api_key: Optional[str],
+        cse_id: Optional[str],
+        **kwargs: Any
+    ) -> List[dict]:
+        service = build("customsearch", "v1", developerKey=api_key)
+        res = service.cse().list(q=search_term, cx=cse_id, **kwargs).execute()
+        return res["items"]
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validate that api key and python package exists in environment."""
+        google_api_key = get_from_dict_or_env(
+            values, "google_api_key", "GOOGLE_API_KEY"
+        )
+        values["google_api_key"] = google_api_key
+
+        google_cse_id = get_from_dict_or_env(values, "google_cse_id", "GOOGLE_CSE_ID")
+        values[google_cse_id] = google_cse_id
+
+        try:
+            from googleapiclient.discovery import build
+
+            build("customsearch", "v1", developerKey=google_api_key)
```

Comment:
this is just to try connecting? could we pass this into init as search engine and use on line 83?

Reply:
I think i fixed, but not sure if I did what you were thinking. Feel free to modify it